File: 101 - GMail API integration/test_scripts/gmail_auth.py
"""Gmail API Authentication Module.

Test script for: Gmail API Authentication
From document: Gmail-API-Integration-Guide.md
Document lines: 868-926

Tests the authentication module from the training guide.
"""
import os
from google.auth.transport.requests import Request
from google.oauth2.credentials import Credentials
from google_auth_oauthlib.flow import InstalledAppFlow
from googleapiclient.discovery import build

# Define your scopes - IMPORTANT: Only request scopes that are configured
# in your OAuth Consent Screen's Data Access section
# For this test, we use only gmail.readonly which is commonly configured
SCOPES = [
    'https://www.googleapis.com/auth/gmail.readonly',
]

def get_gmail_service(credentials_path: str = 'credentials.json',
                      token_path: str = 'token.json'):
    """
    Authenticate and return Gmail API service.

    Args:
        credentials_path: Path to OAuth credentials JSON file
        token_path: Path to store/retrieve access tokens

    Returns:
        Gmail API service instance

    Raises:
        FileNotFoundError: If credentials file is not found
    """
    if not os.path.exists(credentials_path):
        raise FileNotFoundError(
            f"Credentials file not found at {credentials_path}. "
            "Download it from Google Cloud Console."
        )

    creds = None

    # Load existing token if available
    if os.path.exists(token_path):
        creds = Credentials.from_authorized_user_file(token_path, SCOPES)

    # Refresh or create new credentials
    if not creds or not creds.valid:
        if creds and creds.expired and creds.refresh_token:
            creds.refresh(Request())
        else:
            flow = InstalledAppFlow.from_client_secrets_file(
                credentials_path, SCOPES
            )
            creds = flow.run_local_server(port=0)

        # Save credentials for future use
        with open(token_path, 'w') as token:
            token.write(creds.to_json())

    return build('gmail', 'v1', credentials=creds)


if __name__ == '__main__':
    import sys

    # Use credentials from parent directory
    CREDENTIALS_PATH = os.path.join(os.path.dirname(__file__), '..', 'GMailSkill-Credentials.json')
    TOKEN_PATH = os.path.join(os.path.dirname(__file__), '..', 'gmail_token.json')

    print("=" * 60)
    print("Test: Gmail API Authentication Module")
    print("=" * 60)

    # Validate credentials file exists
    if not os.path.exists(CREDENTIALS_PATH):
        print(f"[ERROR] Credentials file not found: {CREDENTIALS_PATH}")
        sys.exit(1)
    print(f"[OK] Credentials file found: {CREDENTIALS_PATH}")

    try:
        # Attempt to authenticate
        print("\nAttempting Gmail API authentication...")
        print("(This may open a browser window for first-time authentication)")

        service = get_gmail_service(
            credentials_path=CREDENTIALS_PATH,
            token_path=TOKEN_PATH
        )

        # Verify by getting profile
        profile = service.users().getProfile(userId='me').execute()

        print(f"\n[OK] Authentication successful!")
        print(f"Email: {profile.get('emailAddress', 'N/A')}")
        print(f"Messages Total: {profile.get('messagesTotal', 'N/A')}")
        print(f"Threads Total: {profile.get('threadsTotal', 'N/A')}")

        print("\n" + "=" * 60)
        print("Test: PASSED")
        print("=" * 60)

    except Exception as e:
        print(f"\n[ERROR] Authentication failed: {e}")
        import traceback
        traceback.print_exc()
        sys.exit(1)

File: 101 - GMail API integration/test_scripts/gmail-auth.js
/**
 * Gmail API Authentication Module for Node.js
 *
 * Test script for: Gmail API Authentication (Node.js)
 * From document: Gmail-API-Integration-Guide.md
 * Document lines: 1727-1811
 */
const fs = require('fs').promises;
const path = require('path');
const { authenticate } = require('@google-cloud/local-auth');
const { google } = require('googleapis');

// Define your scopes - IMPORTANT: Only request scopes configured
// in your OAuth Consent Screen's Data Access section
const SCOPES = [
  'https://www.googleapis.com/auth/gmail.readonly',
];

const TOKEN_PATH = path.join(__dirname, '..', 'gmail_token_nodejs.json');
const CREDENTIALS_PATH = path.join(__dirname, '..', 'GMailSkill-Credentials.json');

/**
 * Reads previously authorized credentials from the save file.
 * @returns {Promise<OAuth2Client|null>}
 */
async function loadSavedCredentialsIfExist() {
  try {
    const content = await fs.readFile(TOKEN_PATH);
    const credentials = JSON.parse(content);
    return google.auth.fromJSON(credentials);
  } catch (err) {
    return null;
  }
}

/**
 * Serializes credentials to a file.
 * @param {OAuth2Client} client
 */
async function saveCredentials(client) {
  const content = await fs.readFile(CREDENTIALS_PATH);
  const keys = JSON.parse(content);
  const key = keys.installed || keys.web;
  const payload = JSON.stringify({
    type: 'authorized_user',
    client_id: key.client_id,
    client_secret: key.client_secret,
    refresh_token: client.credentials.refresh_token,
  });
  await fs.writeFile(TOKEN_PATH, payload);
}

/**
 * Load or request authorization to call APIs.
 * @returns {Promise<OAuth2Client>}
 */
async function authorize() {
  let client = await loadSavedCredentialsIfExist();
  if (client) {
    return client;
  }
  client = await authenticate({
    scopes: SCOPES,
    keyfilePath: CREDENTIALS_PATH,
  });
  if (client.credentials) {
    await saveCredentials(client);
  }
  return client;
}

/**
 * Get Gmail API service instance.
 * @returns {Promise<gmail_v1.Gmail>}
 */
async function getGmailService() {
  const auth = await authorize();
  return google.gmail({ version: 'v1', auth });
}

module.exports = {
  authorize,
  getGmailService,
  SCOPES,
};

// Test execution
if (require.main === module) {
  (async () => {
    console.log('='.repeat(60));
    console.log('Test: Gmail API Authentication Module (Node.js)');
    console.log('='.repeat(60));

    try {
      // Check credentials file exists
      try {
        await fs.access(CREDENTIALS_PATH);
        console.log(`[OK] Credentials file found: ${CREDENTIALS_PATH}`);
      } catch {
        console.log(`[ERROR] Credentials file not found: ${CREDENTIALS_PATH}`);
        process.exit(1);
      }

      console.log('\nAttempting Gmail API authentication...');
      console.log('(This may open a browser window for first-time authentication)');

      const gmail = await getGmailService();

      // Verify by getting profile
      const profile = await gmail.users.getProfile({ userId: 'me' });

      console.log('\n[OK] Authentication successful!');
      console.log(`Email: ${profile.data.emailAddress}`);
      console.log(`Messages Total: ${profile.data.messagesTotal}`);
      console.log(`Threads Total: ${profile.data.threadsTotal}`);

      console.log('\n' + '='.repeat(60));
      console.log('Test: PASSED');
      console.log('='.repeat(60));
    } catch (error) {
      console.log(`\n[ERROR] Authentication failed: ${error.message}`);
      console.error(error);
      process.exit(1);
    }
  })();
}
